pkg/services/ngalert/state/manager.go

```diff
@@ -320,6 +316,47 @@ func (st *Manager) saveAlertStates(ctx context.Context, logger log.Logger, state
 	}
 }
 
+func (st *Manager) logStateTransitions(ctx context.Context, alertRule *ngModels.AlertRule, newStates, staleStates []StateTransition) {
```

Comment:
We can combine `newStates` and `staleStates` into a single slice if the deletion handler only returns resolved states. This is because `s.Resolved: true` implies `s.PreviousState == eval.Alerting && s.State == eval.Normal`, so `s.changed()` will be true and we'll be guaranteed to not accidentally include any pending/other states that got aged out. We still only need one database call this way.

Reply:
`Change` method does not work here because it returns true to `Normal (no-reason) -> Normal (stale state reason)`

Reply:
Yes, I agree it doesn't work here as defined currently. I am saying if the list only contained resolved states, it would work because of the guarantees that Resolved gives us.